keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,1073 @@
+
+### Namespace policy update warnings
+
+When an `enforce` policy (or version) label is added or changed, the admission plugin will test each pod
+in the namespace against the new policy. Violations are returned to the user as warnings. These
+checks have a timeout of XX seconds and a limit of YY pods, and will return a warning in the event
+that not every pod was checked. User exemptions are ignored by these checks, but runtime class
+exemptions still apply. Namespace exemptions are also ignored, but an additional warning will be
+returned when updating the policy on an exempt namespace. These checks only consider actual Pod
+resources, not [templated pods].
+
+These checks are also performed when making a dry-run request, which can be an effective way of
+checking for breakages before updating a policy, for example:
+
+```
+kubectl label --dry-run=server --overwrite ns --all pod-security.kubernetes.io/enforce=baseline
+```
+
+<<[UNRESOLVED]>>
+
+_Non-blocking: can be decided on the implementing PR_
+
+- What should the timeout be for pod update warnings?
+  - Total is a parameter on the context (query parameter for webhooks). Cap should be
+    `min(timeout_param, hard_cap)`, where the `hard_cap` is a small number of seconds.
+  - Expect evaluation to be fast, so even 3k pods should come in well under the timeout.
+- What should the pod limit be set to?
+  - 3,000 is the
+    [documented](https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md)
+    scalability limit for per-namespace pod count.
+  - Warnings should be aggregated for large namespaces (soft cap number of warnings, hard cap number
+    of evaluations).
+
+<<[/UNRESOLVED]>>
+
+
+### Admission Configuration
+
+A number of options can be statically configured through the [Admission Configuration file][]:
+
+```
+apiVersion: apiserver.config.k8s.io/v1
+kind: AdmissionConfiguration
+plugins:
+- name: PodSecurity
+  configuration:
+    defaults:  # Defaults applied when a mode label is not set.
+      enforce:         <default enforce policy level>
+      enforce-version: <default enforce policy version>
+      audit:         <default audit policy level>
+      audit-version: <default audit policy version>
+      warn:          <default warn policy level>
+      warn-version:  <default warn policy version>
+    exemptions:
+      usernames:         [ <array of authenticated usernames to exempt> ]
+      runtimeClassNames: [ <array of runtime class names to exempt> ]
+      namespaces:        [ <array of namespaces to exempt> ]
+...
+```
+
+[Admission Configuration file]: https://github.com/kubernetes/kubernetes/blob/3d6026499b674020b4f8eec11f0b8a860a330d8a/staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1/types.go#L27
+
+#### Defaulting
+
+The default policy level and version for each mode (when no label is present) can be statically
+configured. The default for the static configuration is `privileged` and `latest`.
```

Comment:
> "having the admission plugin on and effectively not-enforcing anything is better than having it off"?

Yes, this is the motivation to me. I would like this plugin (or equivalent) to be almost universally on, and if we do any enforcement by default it will make that harder. The section "Rollout of baseline-by-default for unlabeled namespaces" proposes a possible path to baseline-by-default _after_ this feature goes to GA.